pysingfel/detector/autoranging.py

```diff
@@ -10,54 +10,55 @@ def __init__(self, cameraConfig, *args, **kwargs):
         self.cameraConfig = cameraConfig
 
     def updateFlatGains(self, flatGains):  ## not sure this makes sense in the residual world
-        self.gains = self.matricize(flatGains)
+        self.gains = self.smear(self.matricize(flatGains), None, None)
 
     def updateFlatOffsets(self, flatOffsets):
-        self.offsets = self.matricize(flatOffsets)
+        self.offsets = self.smear(self.matricize(flatOffsets), None, None)
 
     def updateFlatSwitchPoints(self, flatSwitchPoints):
-        self.switchPoints = self.matricize(flatSwitchPoints)
+        self.switchPoints = self.smear(self.matricize(flatSwitchPoints), None, None)
 
     def updateSwitchPoints(self, switchPoints):
         self.switchPoints = switchPoints
 
     def setupMatrices(self):
-        self.residualGains = self.matricize(self.gains, self.gainErrors, None)/self.matricize(self.gains, None, None) # Q: ADU/keV?
+        self.residualGains = self.smear(self.matricize(self.gains), self.gainErrors, None)/self.smear(self.matricize(self.gains), None, None) # Q: ADU/keV?
```

Comment:
Inefficient to call matricize(gains) twice here.
Actually, the denominator is wrong. It should just be matricize()?
Please check whether the output is still the same as the original code.

Reply:
I fixed the particle orientation and printed out the calibrated photons at pixel[10][0][383].
('calib_photons[10][0][383]: ', 521.7391304347826) --> The output is still the same as the original code.
<img width="555" alt="Screen Shot 2020-07-17 at 7 34 22 PM" src="https://user-images.githubusercontent.com/23045561/87842865-a47a8880-c864-11ea-9b65-092a7e6b2072.png">

Reply:
The return value of `matricize()` is tmp (type:list, len=nRanges). Pass it into `smear` and apply relativeSmear/absoluteSmear then the return value is np.array(tmp) (type:array, size=detector dimension).

Reply:
matrice() wasn't returning an array but a list

Reply:
I ran the example script `ExampleAutoranging.py` with the updated `autoranging.py` but got the following error:
(ana-1.5.12) [iris@psanagpu109 scripts]$ python ExampleAutoranging.py 
/reg/neh/home/iris/Software/pysingfel/pysingfel/detector/autoranging.py:25: RuntimeWarning: invalid value encountered in divide
  self.residualGains = self.smear(self.matricize(self.gains), self.gainErrors, None)/self.matricize(self.gains) # unitless
/reg/neh/home/iris/Software/pysingfel/pysingfel/build_autoranging_frames.py:43: RuntimeWarning: invalid value encountered in greater
  self.frame[np.where(self.frame > self.det.maxEnergy)] = self.det.maxEnergy
('calib_photons[10][0][383]: ', nan)
Traceback (most recent call last):
  File "ExampleAutoranging.py", line 48, in <module>
    viz.imshow(img)
  File "/reg/neh/home/iris/Software/pysingfel/pysingfel/visualizer.py", line 47, in imshow
    plt.colorbar()
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/pyplot.py", line 2320, in colorbar
    ret = gcf().colorbar(mappable, cax = cax, ax=ax, **kw)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/figure.py", line 2098, in colorbar
    cb = cbar.colorbar_factory(cax, mappable, **cb_kw)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 1399, in colorbar_factory
    cb = Colorbar(cax, mappable, **kwargs)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 945, in __init__
    ColorbarBase.__init__(self, ax, **kw)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 327, in __init__
    self.draw_all()
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 353, in draw_all
    self._config_axes(X, Y)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 461, in _config_axes
    self.update_ticks()
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 378, in update_ticks
    ticks, ticklabels, offset_string = self._ticker()
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/colorbar.py", line 622, in _ticker
    b = np.array(locator())
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/ticker.py", line 2098, in __call__
    return self.tick_values(vmin, vmax)
  File "/reg/g/psdm/sw/conda/inst/miniconda2-prod-rhel7/envs/ana-1.5.12/lib/python2.7/site-packages/matplotlib/ticker.py", line 2124, in tick_values
    "Data has no positive values, and therefore can not be "
ValueError: Data has no positive values, and therefore can not be log-scaled.